Remove MAC lookup for minecraft server

girlpunk opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Minecraft servers no longer rely on the MAC address of the server as a unique identifier when using an IP address to access the host.

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adding a Minecraft server will fail if the server uses an IP address but is not on the same layer 2 network. Local addresses and DNS names work fine, this functionality has not changed.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #33955
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

girlpunk committed:
Remove MAC lookup for minecraft server

This will fail if the MC server uses an IP address but is not on the same layer 2 network.
Local addresses and DNS names work fine, as found in bug #33955

girlpunk committed:
Fix CI warnings

girlpunk committed:
Remove unit test for IP validation

This is no longer needed, the component does not check that the host is always valid, as DNS names and SRV records are not IP addresses but are still acceptable.

girlpunk committed:
Update test to match new expected title

MartinHjelmare commented on homeassistant/components/minecraft_server/config_flow.py:
```diff
@@ -92,26 +64,16 @@ class MinecraftServerConfigFlow(ConfigFlow, domain=DOMAIN):
-                            unique_id = f"{host}-{port}"
+                        # Use host name and port in unique_id (to support servers with
+                        # same host name but different ports).
+                        unique_id = f"{host}-{port}"
```
What kind of host address is this? Could it change?

cyberjacob replied:
The host is specified in the configuration, it may be an IPv4/IPv6 address or a DNS name. For example: `minecraft.example.com` or `1.1.1.1`.

MartinHjelmare replied:
We normally don't allow IP addresses as unique id as they may change for the same device or service.

elmurato replied:
That is why in the initial PR for the integration I used the MAC address as the unique ID in case of IP addresses (with the limitation that it doesn't work if the MAC address is not readable).

Instead, a second step in the config flow could be added for manual entry of the MAC address, if it's not readable...

elmurato replied:
Found the mentioned review finding:
https://github.com/home-assistant/core/pull/30992#discussion_r373869101

cyberjacob replied:
If your argument is that the IP address may change, the same could happen for a MAC address. And that a different IP could easily be considered a different server. MAC addresses are also shared between different IPs on a single host, causing a possibility for conflicts.

I'd also point out that for the identifier to change, the server would have to be manually removed and re-added by the user. Servers accessed via a DNS record (weather A, AAAA, or SRV) use this record as part of the identifier (this functionality has not changed).

MartinHjelmare replied:
@balloob?

balloob replied:
No IP addresses.

cyberjacob replied:
Ok, do you have an alternate suggestion for fixing this bug?

elmurato replied:
As I said before: One solution would be to ask the user for the MAC address, in case it can‘t be read. This can be probably done with a 2nd step in the config flow.

TheDevFreak replied:
> We normally don't allow IP addresses as unique id as they may change for the same device or service.

If the case that the IP address changes the integration would have to be removed and set up again anyway since the Minecraft server would no longer exist on the IP address that Home Assistant would be querying anyway - Because of this I can't see any problem with using the IP address as the entity id? Failing that being allowed I can't see any other attribute that could be the unique ID, aside from the MAC address - which breaks under reasons given above.

tigattack replied:
What if I were to have two Minecraft servers on the same host, and therefore same MAC address?  
Until very recently this was the case for me; I had one server on the default port (25565) and another on 25566. From what I can see, this integration would not be able to handle that since it's relying on the MAC address as a unique ID.  
As pointed out above, if the IP of the server changed then the integration would break anyway, resulting in the need to configure it from scratch, which would in turn result in a new unique ID for the entities.

t3rminalVel0city replied:
Surely a MAC address isn't really a unique ID? If I'm running more than one minecraft server on the same host, or using a shared hosting platform with multiple servers per host then more than one minecraft server could share the same MAC address. In this instance, an IP:Port combination would be preferred, as it's generally not possible to run multiple services on the same port on a host. 

Additionally, generally servers have static IP assignments and thus the IP address will be unlikely to change without specific user intervention and even if the IP did change, the integration would have to be reconfigured anyway as it connects to an IP and port in order to read data from the minecraft instance.

balloob replied:
Good point. We should remove mac address too.

Use config entry ID.

girlpunk committed:
Retain square brackets

To avoid NRI conflicts (https://en.wikipedia.org/wiki/IPv6_address#Literal_IPv6_addresses_in_network_resource_identifiers)

girlpunk committed:
Correct tests for IPv6 address with brackets

firestrife23 commented:
Why cancelled?? I cannot add Minecraft integration because I'm running Minecraft in Docker, and using IP address is far more reliable than MAC address lookup.

cyberjacob commented:
@firestrife23 Cancelled because the HASS team refuse to accept the changes. Take a look at the fork of this integration at https://github.com/cyberjacob/minecraft-home-assistant/ which includes the changes.

frenck commented:
> Why cancelled?? I cannot add Minecraft integration because I'm running Minecraft in Docker, and using IP address is far more reliable than MAC address lookup.

Using IP address for an unique ID is just wrong and should not be done. It will cause problems in the long run. Instead, don't use a unique ID or use the one provided by the config entry.

> Cancelled because the HASS team refuse to accept the changes.

That is incorrect, the PR was closed because the PR author hasn't implemented the needed requirements and left it running stale.

firestrife23 commented:
> Using IP address for an unique ID is just wrong and should not be done. It will cause problems in the long run. Instead, don't use a unique ID or use the one provided by the config entry.

While that may be true for some people, however, my servers are assigned with static IP. Thank you for clarification, and I will use @cyberjacob's fork in the meantime.
